### DISCUSSION TITLE
Establishing whether my exported content contains colour

### DISCUSSION THREAD

**Discussion by AVPGEC** [ImageMagick/ImageMagick#4952] [Help] [answered]
Hi,

I am a start up graphic designer and I am trying to send some content to a professional printer. Whilst some of my content is colour there are some images that I am happy to print using black only. The reason for this is the printer will charge me less for those images printed using only black. The printer is using a tool (I believe part of the Adobe toolset) to establish whether or not the pdf I am sending contains colour and they are telling me it does despite me exporting to pdf in grayscale from GIMP.

I have just downloaded the latest version of ImageMagick as I read some posts that implied it could give me a definitive answer as to whether or not my image contains any colours however I am struggling to get any of the previous solutions to work. I would be grateful if anyone could provide me with a simple instruction that I could follow that would allow me to interrogate my content for whether or not it contains any colour?

Thanks in advance

**Comment by fmw42** [comment]
Perhaps your printer tool is just looking for a 3 channel image vs a 1 channel image? If your grayscale image is RGB 3 equal channels, it might be telling your that it is color. Is your PDF a raster image or just vector data? If a raster, what format is the raster that is inside the PDF vector shell?  Can you post an example PDF that is giving your issues?

**Reply by AVPGEC** [reply]
It would be a raster image.

Examples attached of the GIMP file exported as jpg and as pdf. The native GIMP file (xcf) isn't accepted as a file type upload here.

[example pdf.pdf](https://github.com/ImageMagick/ImageMagick/files/8255926/example.pdf.pdf)

[example jpg](https://user-images.githubusercontent.com/101661500/158439371-f7113a9a-7a06-4c86-937b-33f19692c27e.jpg)

**Comment by fmw42** [comment, marked as answer]
I try to convert your image to grayscale.

`convert example.pdf.pdf -colorspace gray -type grayscalealpha new_example.pdf`

[new_example.pdf](https://github.com/ImageMagick/ImageMagick/files/8255998/new_example.pdf)

However, before and after IM verbose information says it colorspace RGB and type grayscale alpha. So no difference. 

However, using EXIFTOOL, I do see one difference.

The new file has a tag:

`- EmbeddedImageColorSpace         : DeviceGray`

That is not in your original file.

So check if my new pdf works or not on your printer. If it works, then this tag may be the issue that is missing in your GIMP conversion.

**Reply by AVPGEC** [reply]
Thank you, I will try that and report back.

**Reply by AVPGEC** [reply]
Hi,

The printer has confirmed this passes whatever interrogation they are performing to check it is not colour. Thank you.

Obviously I need to be able to repeat this result myself. Can I check that the instruction you gave above, i.e.

convert example.pdf.pdf -colorspace gray -type grayscalealpha new_example.pdf

was correct? I'm asking because there are two problems with it that I'm experiencing: 1) The file name I gave you wasn't example.pdf.pdf, it was example pdf.pdf. I could understand if you had replaced the space with an underscore but another . doesn't make sense to me? 2) When I try to do this I am getting an error message. I can convert the jpg that I attached as an example to a pdf but I cannot convert a pdf. Can you confirm that it was the original example pdf you converted using Magick?

**Reply by fmw42** [reply]
That is the correct code. The important part is (for images with transparency)

`-colorspace gray -type grayscalealpha`

You can name the output anything you want or keep the same input name and write over it. If you cannot read in a PDF, then either you do not have Ghostscript installed or your policy.xml file has restricted reading of PDF files.  For the former, try

`convert -version`

and see if gs or gslib is listed in the Delegates line.

For the second, you would need to edit your policy.xml file to give read|write permissions for the line for PS/PDF/EPS etc. So change

`<policy domain="coder" rights="write" pattern="{MSVG,MVG,PS,PDF,XPS}" />`
  
to

`<policy domain="coder" rights="read|write" pattern="{MSVG,MVG,PS,PDF,XPS}" />`
 
see https://imagemagick.org/source/policy.xml and https://imagemagick.org/script/resources.php

**Reply by AVPGEC** [reply]
Thanks for that. As this is only day 2 of me using ImageMagick I'm not going to pretend my understanding of everything is complete but installing Ghostscript has resolved this for me (I thought I already had it installed but it must have been a previous Windows installation). I can now convert and identify pdfs with ImageMagick.

I'm now going to a look at EXIFTOOL and see if I can replicate seeing the difference in tags that you are able to see with the pdf I've converted myself. If I can then I'll check the pdf I have converted myself with the printer one final time just to make 100% sure I can replicate your result and then I think my issue is resolved.

Thank you once again for your help. If the above all goes as I hope then I'll make sure I come back on here and mark your answer as the answer.

I'm curious as to whether it is possible for me to get the same end result natively out of GIMP (I have tried a couple of the more obvious methods) without having to add the additional step of using ImageMagick but I guess if the tag that can be seen in EXIFTOOL is the key attribute then that is something I can use as a basis to test and explore with GIMP outputs in the longer term.

**Reply by fmw42** [reply]
Here is what I did:

Your file:

```
exiftool -s -ee -g1 -u -n -D example.pdf.pdf
---- ExifTool ----
    - ExifToolVersion                 : 10.51
---- System ----
    - FileName                        : example.pdf.pdf
    - Directory                       : .
    - FileSize                        : 5483
    - FileModifyDate                  : 2022:03:15 10:51:35-07:00
    - FileAccessDate                  : 2022:03:16 09:46:57-07:00
    - FileInodeChangeDate             : 2022:03:15 10:51:35-07:00
    - FilePermissions                 : 644
---- File ----
    - FileType                        : PDF
    - FileTypeExtension               : PDF
    - MIMEType                        : application/pdf
---- PDF ----
    - PDFVersion                      : 1.5
    - Linearized                      : false
    - PageCount                       : 1
    - Producer                        : cairo 1.17.4 (https://cairographics.org)
    - CreateDate                      : 2022:03:15 17:38:58Z

```

My converted file with ImageMagick:

```
exiftool -s -ee -g1 -u -n -D new_example.pdf
---- ExifTool ----
    - ExifToolVersion                 : 10.51
---- System ----
    - FileName                        : new_example.pdf
    - Directory                       : .
    - FileSize                        : 15026
    - FileModifyDate                  : 2022:03:15 10:53:33-07:00
    - FileAccessDate                  : 2022:03:16 09:48:11-07:00
    - FileInodeChangeDate             : 2022:03:15 10:53:33-07:00
    - FilePermissions                 : 644
---- File ----
    - FileType                        : PDF
    - FileTypeExtension               : PDF
    - MIMEType                        : application/pdf
    - FileType                        : (unsupported)
---- PDF ----
    - PDFVersion                      : 1.4
    - Linearized                      : false
    - Title                           : new_example
    - CreateDate                      : 2022:03:15 17:53:33
    - ModifyDate                      : 2022:03:15 17:53:33
    - Producer                        : https://imagemagick.org
    - EmbeddedImageFilter             : RunLengthDecode
    - EmbeddedImageWidth              : 737
    - EmbeddedImageHeight             : 522
    - EmbeddedImageColorSpace         : DeviceGray  <----- See here
    - PageCount                       : 1
```

**Reply by AVPGEC** [reply]
Thanks for that, that has saved a lot of time. I've downloaded EXIFTOOL and can see the exact same tag in the pdf I converted so it looks promising - time to confirm with the printer.

**Comment by snibgo** [comment]
The tag can be set with exiftool. That may be a better solution than using IM to read a PDF and write a PDF.

**Reply by AVPGEC** [reply]
Interesting. Are you able to provide an instruction for how I can do that? I'm sending the email to the printer tonight so if it possible to do that then I'd also include a second copy of the pdf with the tag simply added by EXIFTOOL and see if the printer is still satisfied that their software is telling them it can be printed using black only.

**Comment by snibgo** [comment]
Like this, but it doesn't seem to work:
```
f:\web\im>exiftool -PDF:EmbeddedImageColorSpace=DeviceGray t.pdf

Warning: Sorry, PDF:EmbeddedImageColorSpace doesn't exist or isn't writable
Nothing to do.
```

**Reply by AVPGEC** [reply]
Yes, I get the same error message. I'll stick to using ImageMagick now and trying to do it natively from GIMP in the longer term, but thanks for trying.
